File: mergesort.c
#include<stdio.h>

void merge(int arr[100],int lb,int mid,int ub)      //merge function
{
	int i , k , j;
	int b[100];
	i=lb;
	j=mid+1;
	k=lb;
	
	while((i<=mid && j<=ub))
	{
		if(arr[i]<=arr[j])
		{
			b[k]=arr[i];
			i++;
			k++;
		}
		else
		{
			b[k]=arr[j];
			j++;
			k++;
		}
	}
	while(j<=ub)
		{
			b[k]=arr[j];
			j++;
			k++;
		}
	

	while(i<=mid)
		{
			b[k]=arr[i];
			i++;
			k++;
		}
	for(k=lb;k<=ub;k++)
	{
		arr[k]=b[k];
	}
}

void merge_sort(int arr[100],int lb, int ub)             //function to sort
{
	int mid;
	if(lb<ub)
	{
		mid=(lb+ub)/2;
		merge_sort(arr,lb,mid);
		merge_sort(arr,mid+1,ub);
		merge(arr,lb,mid,ub);	
	}
}

void print(int arr[100],int n)                  //display function
{
	int i ;
	for(i=0;i<n;i++)
	{
		printf(" %d ",arr[i]);
    } 
}

int main()
{
	int arr[100];
	int n ,i;
	printf("Enter the number of array elemts:");
	scanf("%d",&n);
	printf("Enter the array elemnets:\n");
	for(i=0;i<n;i++)
	{
		scanf("%d",&arr[i]);
	}
	printf("Before sorting the array elements are:\n");
	print(arr,n);
	merge_sort(arr,0,n-1);
	printf("\nAfter sorting the array elements are:\n");
	print(arr,n);	
}

File: Mergesort.java
import java.util.*;

public class Main
{   
    public static void conquer(int arr[],int si ,int mid ,int ei)
    {
        int merge[]= new int[ei-si+1];         // new Array
        int index1=si;       //track 1st Array
        int index2=mid+1;    //track 2nd Array
        
        int x=0;          // track the index of new merge Array
        
        while(index1<=mid && index2<=ei)
        {
            if(arr[index1]<=arr[index2])
            {
                merge[x]=arr[index1];
                x++;
                index1++;
            }
            else
            {
                merge[x++]=arr[index2++];  
            }
        }
        
        while(index1<=mid)
        {
            merge[x++]=arr[index1++];
        }
        
        while(index2<=ei)
        {
            merge[x++]=arr[index2++];
        }
        
        for(int i=0,  j=si; i<merge.length;i++, j++)            // copy form new array -> original array
        {
            arr[j]=merge[i];
        }
    }
    
    
    
    
    // function for Divide

    public static void Divide(int arr[],int si, int ei)        // (Array , starting Index , end index)
    {
       if(si>=ei)           // when single element left
       { 
           return;
       }
       
        int mid = (si+(ei-si)/2);                   //   means (si+ei)/2
        Divide(arr,si,mid);        //recursive call
        Divide(arr,mid+1,ei);
        
        conquer(arr,si,mid,ei);
        
        
    }
    
    // function for display
    
    public static void display(int arr[])
    {
        int n = arr.length;
        
        for(int i=0;i<n;i++)
        {
            System.out.print(arr[i]+" ");  
        }
        
        System.out.println();
    }
    
    
    
    public static void main(String[] args) {
        
        
        
        
        Scanner scan = new Scanner(System.in);
        
		System.out.println("Enter the length of the array: ");
		int n= scan.nextInt();
		
		int arr[]= new int[n];
		
		System.out.println("Enter the array elements: ");
		
		for(int i=0;i<n;i++)
		{
		    arr[i]= scan.nextInt();
		}
		
		System.out.println("Elements before Sorting: ");
		display(arr);
		Divide(arr,0,n-1);
		System.out.println("Elements after Sorting: ");
		display(arr);

		
	}
    
    
}
